src/ucm/cuda/cudamem.c

```diff
@@ -6,392 +6,263 @@
+                    size_t, unsigned)
+UCM_CUDA_FREE_FUNC(cuMemFree, CUresult, CUdeviceptr, UCS_MEMORY_TYPE_CUDA)
+UCM_CUDA_FREE_FUNC(cuMemFree_v2, CUresult, CUdeviceptr, UCS_MEMORY_TYPE_CUDA)
+UCM_CUDA_FREE_FUNC(cuMemFreeHost, CUresult, void*, UCS_MEMORY_TYPE_HOST)
+UCM_CUDA_FREE_FUNC(cuMemFreeHost_v2, CUresult, void*, UCS_MEMORY_TYPE_HOST)
+
+static ucm_cuda_func_t ucm_cuda_driver_funcs[] = {
+    UCM_CUDA_FUNC_ENTRY(cuMemAlloc),
+    UCM_CUDA_FUNC_ENTRY(cuMemAlloc_v2),
+    UCM_CUDA_FUNC_ENTRY(cuMemAllocManaged),
+    UCM_CUDA_FUNC_ENTRY(cuMemAllocPitch),
+    UCM_CUDA_FUNC_ENTRY(cuMemAllocPitch_v2),
+    UCM_CUDA_FUNC_ENTRY(cuMemFree),
+    UCM_CUDA_FUNC_ENTRY(cuMemFree_v2),
+    UCM_CUDA_FUNC_ENTRY(cuMemFreeHost),
+    UCM_CUDA_FUNC_ENTRY(cuMemFreeHost_v2),
+    {{NULL}, NULL}
+};
 
-    return ret;
-}
+/* Runtime API replacements */
+UCM_CUDA_ALLOC_FUNC(cudaMalloc, UCS_MEMORY_TYPE_CUDA, cudaError_t, cudaSuccess,
+                    arg0, void*, "size=%zu", size_t)
+UCM_CUDA_ALLOC_FUNC(cudaMallocManaged, UCS_MEMORY_TYPE_CUDA_MANAGED,
+                    cudaError_t, cudaSuccess, arg0, void*,
+                    "size=%zu flags=0x%x", size_t, unsigned)
+UCM_CUDA_ALLOC_FUNC(cudaMallocPitch, UCS_MEMORY_TYPE_CUDA, cudaError_t,
+                    cudaSuccess, (size_t)arg1 * arg2, void*,
+                    "pitch=%p width=%zu height=%zu", size_t*, size_t, size_t)
+UCM_CUDA_FREE_FUNC(cudaFree, cudaError_t, void*, UCS_MEMORY_TYPE_CUDA)
+UCM_CUDA_FREE_FUNC(cudaFreeHost, cudaError_t, void*, UCS_MEMORY_TYPE_HOST)
+
+static ucm_cuda_func_t ucm_cuda_runtime_funcs[] = {
+    UCM_CUDA_FUNC_ENTRY(cudaFree),
+    UCM_CUDA_FUNC_ENTRY(cudaFreeHost),
+    UCM_CUDA_FUNC_ENTRY(cudaMalloc),
+    UCM_CUDA_FUNC_ENTRY(cudaMallocManaged),
+    UCM_CUDA_FUNC_ENTRY(cudaMallocPitch),
+    {{NULL}, NULL}
+};
 
-cudaError_t ucm_cudaMallocPitch(void **devPtr, size_t *pitch,
-                                size_t width, size_t height)
+static ucs_status_t
+ucm_cuda_install_hooks(ucm_cuda_func_t *funcs, int *used_reloc)
 {
-    cudaError_t ret;
-
-    ucm_event_enter();
-
-    ret = ucm_orig_cudaMallocPitch(devPtr, pitch, width, height);
-    if (ret == cudaSuccess) {
-        ucm_trace("ucm_cudaMallocPitch(devPtr=%p size:%lu)",*devPtr, (width * height));
-        ucm_dispatch_mem_type_alloc(*devPtr, (width * height), UCS_MEMORY_TYPE_CUDA);
-        ucm_cuda_set_ptr_attr((CUdeviceptr) *devPtr);
-    }
+    ucm_cuda_func_t *func;
+    ucs_status_t status;
+    void *func_ptr;
+
+    *used_reloc = 0;
+    for (func = funcs; func->patch.symbol != NULL; ++func) {
+        func_ptr = ucm_reloc_get_orig(func->patch.symbol, func->patch.value);
+        if (func_ptr == NULL) {
+            continue;
+        }
 
-    ucm_event_leave();
-    return ret;
-}
+        status = UCS_ERR_UNSUPPORTED;
 
-cudaError_t ucm_cudaHostGetDevicePointer(void **pDevice, void *pHost, unsigned int flags)
-{
-    cudaError_t ret;
+        if (ucm_mmap_hook_mode() == UCM_MMAP_HOOK_BISTRO) {
```

Comment:
@yosefe  should we detach cuda hooks from ucm_mmap_hook_mode() flag? may be we have a case where we need to to disable malloc bistro hooks but not cuda hooks ( for ex: on summit with spectrum mpi).
how about controlling cuda hooks with new flag "ucm_cuda_hook_mode: and "remove ucm_global_opts.enable_cuda_reloc" ?